Adapter/Algolia.php

```diff
@@ -90,90 +108,234 @@ public function __construct(
         $this->config = $config;
         $this->catalogSearchHelper = $catalogSearchHelper;
         $this->storeManager = $storeManager;
+        $this->registry = $registry;
+        $this->customerSession = $customerSession;
         $this->algoliaHelper = $algoliaHelper;
+        $this->productHelper = $productHelper;
         $this->request = $request;
         $this->documentFactory = $documentFactory;
     }
 
     /**
      * {@inheritdoc}
-     *
-     * @uses getAlgoliaDocument
      */
     public function query(RequestInterface $request)
     {
-        $useNative = false;
         $storeId = $this->storeManager->getStore()->getId();
-        $query = $this->catalogSearchHelper->getEscapedQueryText();
-        $temporaryStorage = $this->temporaryStorageFactory->create();
-        $documents = [];
-        $table = null;
 
-        if ($this->isAllowed($storeId)
-            && ($this->isSearch() ||
+        if (!$this->isAllowed($storeId)
+            || !($this->isSearch() ||
                 $this->isReplaceCategory($storeId) ||
                 $this->isReplaceAdvancedSearch($storeId))
         ) {
-            try {
-                $algoliaQuery = $query !== '__empty__' ? $query : '';
+            return $this->nativeQuery($request);
+        }
 
-                // If instant search is on, do not make a search query unless SEO request is set to 'Yes'
-                if (!$this->config->isInstantEnabled($storeId) || $this->config->makeSeoRequest($storeId)) {
-                    $documents = $this->algoliaHelper->getSearchResult($algoliaQuery, $storeId);
-                }
+        $query = $this->catalogSearchHelper->getEscapedQueryText();
+        $temporaryStorage = $this->temporaryStorageFactory->create();
+        $documents = [];
+        $table = null;
 
-                $apiDocuments = array_map([$this, 'getAlgoliaDocument'], $documents);
-                $table = $temporaryStorage->storeApiDocuments($apiDocuments);
-            } catch (AlgoliaConnectionException $e) {
-                $useNative = true;
+        try {
+            // If instant search is on, do not make a search query unless SEO request is set to 'Yes'
+            if (!$this->config->isInstantEnabled($storeId) || $this->config->makeSeoRequest($storeId)) {
+                $algoliaQuery = $query !== '__empty__' ? $query : '';
+                $documents = $this->getDocumentsFromAlgolia($algoliaQuery, $storeId);
             }
-        } else {
-            $useNative = true;
-        }
 
-        if ($useNative) {
-            $nativeQueryData = $this->getNativeQueryData($request);
-            $documents = $nativeQueryData['documents'];
-            $table = $nativeQueryData['table'];
+            $apiDocuments = array_map([$this, 'getApiDocument'], $documents);
+            $table = $temporaryStorage->storeApiDocuments($apiDocuments);
+        } catch (AlgoliaConnectionException $e) {
+            $this->nativeQuery($request);
```

Comment:
I don't think so ... From what I understand, initially the goal was to use the native behaviour as fallback in case of error (?) : https://github.com/algolia/algoliasearch-magento-2/blob/master/Adapter/Algolia.php#L135

Reply:
I get that, but on [line 131](https://github.com/algolia/algoliasearch-magento-2/pull/616/files/a0a5792df5cf8e094c1acd4b664203a77f21eba3#diff-83e13deff79b331f6a3ca0e7b1a6445aR131) there is `return $this->nativeQuery($request);` instead of just `$this->nativeQuery($request);`.
And the `return` makes sense as you don't want to probably continue in the method as the method is working with Algolia related data.

Reply:
Oh you're right, sorry ;)